File: my-folder/0051-n-queens/solution.cpp
class Solution {
public:
    bool ispos(vector<string> temp, int row,int col,int n){
        int dr = row;
        int dc = col;
        while(row>=0 && col>=0){
            if(temp[row][col]=='Q') return false;
            row--;
            col--;
        }
        row = dr;
        col = dc;
        while(col>=0){
            if(temp[row][col]=='Q') return false;
            col--;
        }
        row = dr;
        col = dc;
        while(row<n && col>=0){
            if(temp[row][col]=='Q') return false;
            row++;
            col--;
        }
        return true;
    }
    void solve(vector<vector<string>> &ans,vector<string> &temp,int n,int col){
        if(col == n){
            ans.push_back(temp);
            return;
        }
        for(int row=0;row<n;row++){
            if(ispos(temp,row,col,n)){
                temp[row][col]='Q';
                solve(ans,temp,n,col+1);
                temp[row][col]='.';
            }

        }

    }
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> ans;
        string s(n,'.');
        vector<string> temp(n,s);
        solve(ans,temp,n,0);
        return ans;
        
    }
};

File: my-folder/0051-n-queens/solution.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> solutions = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = '.';
            }
        }
        solve(solutions, board, 0, n);
        return solutions;
    }

    private void solve(List<List<String>> solutions, char[][] board, int row, int n) {
        if (row == n) {
            solutions.add(createBoard(board, n));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isSafe(board, row, col, n)) {
                board[row][col] = 'Q';
                solve(solutions, board, row + 1, n);
                board[row][col] = '.';
            }
        }
    }

    private boolean isSafe(char[][] board, int row, int col, int n) {
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }
        return true;
    }

    private List<String> createBoard(char[][] board, int n) {
        List<String> res = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            res.add(new String(board[i]));
        }
        return res;
    }
}
